src/server/replicated_archetypes.rs

```diff
@@ -0,0 +1,260 @@
+use std::mem;
+
+use bevy::{
+    ecs::{
+        archetype::{ArchetypeGeneration, ArchetypeId, Archetypes},
+        component::{ComponentId, StorageType},
+    },
+    prelude::*,
+};
+
+use crate::core::{
+    replication_fns::ComponentFnsId, replication_rules::ReplicationRules, Replication,
+};
+
+/// Cached information about all replicated archetypes.
+#[derive(Deref)]
+pub(crate) struct ReplicatedArchetypes {
+    /// ID of [`Replication`] component.
+    marker_id: ComponentId,
+
+    /// Highest processed archetype ID.
+    generation: ArchetypeGeneration,
+
+    /// Archetypes marked as replicated.
+    #[deref]
+    archetypes: Vec<ReplicatedArchetype>,
+}
+
+impl ReplicatedArchetypes {
+    /// ID of [`Replication`] component.
+    pub(crate) fn marker_id(&self) -> ComponentId {
+        self.marker_id
+    }
+
+    /// Updates internal view of the [`World`]'s replicated archetypes.
+    ///
+    /// If this is not called before querying data, the results may not accurately reflect what is in the world.
+    pub(super) fn update(&mut self, archetypes: &Archetypes, rules: &ReplicationRules) {
+        let old_generation = mem::replace(&mut self.generation, archetypes.generation());
+
+        // Archetypes are never removed, iterate over newly added since the last update.
+        for archetype in archetypes[old_generation..]
+            .iter()
+            .filter(|archetype| archetype.contains(self.marker_id))
+        {
+            let mut replicated_archetype = ReplicatedArchetype::new(archetype.id());
+            for rule in rules.iter().filter(|rule| rule.matches(archetype)) {
+                for &(component_id, fns_id) in &rule.components {
+                    // Since rules are sorted by priority,
+                    // we are inserting only new components that aren't present.
+                    if replicated_archetype
+                        .components
+                        .iter()
+                        .any(|component| component.component_id == component_id)
+                    {
+                        continue;
+                    }
```

Comment:
Are you suggesting to print something like "overlapping detected" once?
I think that it might makes sense to print it each time since user can have some "expected" rules already. But printing each time will require allocation since I need to convert `ComponentId` into names, otherwise the warning won't be useful.

What do you think?

Reply:
Or write a debug message like "skipping a component for a rule with priority N because it already contains the same component with priority M"?